### PR TITLE
Framework-dependent System.Drawing.Common version in NuGet packages

### LINKED ISSUE #1311: Central version management
@Kritner from the CodingBlocks slack (https://www.codingblocks.net/slack) has written blog posts ([2021-10-01](https://blog.kritner.com/2021/10/01/dotnet-centralized-package-versioning/), [2018-10-10](https://blog.kritner.com/2018/10/10/using-configuration-files-to-manage-nuget-package-versions-and-other-stuff/)) discussing how to use a single file to define version of multiple projects and their dependencies, making versioning easier to track and update.

## Current System
ScottPlot version and all its dependency package versions are defined in individual csproj files:
* [ScottPlot.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/ScottPlot/ScottPlot.csproj)
* [ScottPlot.WinForms.NUGET.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/controls/ScottPlot.WinForms/ScottPlot.WinForms.NUGET.csproj)
* [ScottPlot.WPF.NUGET.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/controls/ScottPlot.WPF/ScottPlot.WPF.NUGET.csproj)
* [ScottPlot.Avalonia.NUGET.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/controls/ScottPlot.Avalonia/ScottPlot.Avalonia.NUGET.csproj)

A versioning-bumping script is manually run to increment ScottPlot version project-wide:
* [Versioner/Program.cs](https://github.com/ScottPlot/ScottPlot/blob/master/dev/tools/Versioner/Program.cs)

## Proposed System
* Use a central file to store ScottPlot version
  * This should be used for `Version`, `AssemblyVersion`, and `FileVersion`
  * Notice that the first is `x.x.x` and the other two are `x.x.x.0` (that fourth number is important)
* Use a central file to store versions of all dependencies _that we want constrained_ (not necessary all packages).
  * ⚠️ Let's be careful which version of `System.Drawing.Common` is used in the core packages. We found that it this package can be _upgraded_ in old solutions, but it breaks them such that they cannot be _downgraded_ again (#1004). For this reason we probably want to keep targeting `4.6.1` for the primary library packages, although tests and demo applications may use newer versions.
* Consider whether to use GitHub Actions to automatically bump the version number before building and publishing NuGet packages in the cloud. See [`.github/workflows/ci.yaml`](https://github.com/ScottPlot/ScottPlot/blob/master/.github/workflows/ci.yaml)

## Potential Drawbacks
* Old versions of Visual Studio may be unable to use the graphical NuGet package manager. I'm not sure about how widely supported central package management is.
  * I think this is okay because it's unlikely typical users working in the source code of this project will need to add NuGet packages.

### PR DESCRIPTION
**Pull request by Kritner**
**Purpose:**
* As a result of https://blog.kritner.com/2021/10/01/dotnet-centralized-package-versioning/ and chatting with Scott on slack ;)
* I can't seem to build all of the projects, mostly around the pieces that require framework and/or windows based functionality.  I don't know if this is something you would be able to check off @swharden

image

#1274, Closes #1311

### CODE REVIEW

**Comment by swharden** [comment]
Hi @Kritner, thanks for sharing this! See my notes in #1311

If you remove the draft status of this PR you can confirm the CI/tests pass and I won't merge it until you're ready 👍

**Comment by Kritner** [comment]
#1274, Closes #1311

**Comment by swharden** [comment]
> I can't seem to build all of the projects, mostly around the pieces that require framework and/or windows based functionality

This might make this difficult 😅

We exert a lot of effort to maintain cross-targeted builds across .NET Framework (back to 4.6.1), .NET Standard (2.0), and .NET 5 so the library runs on Windows, MacOS, and Linux.

If `ManagePackageVersionsCentrally` isn't supported with those earlier targets, a less fancy solution [like csproj variables](https://blog.kritner.com/2018/10/10/using-configuration-files-to-manage-nuget-package-versions-and-other-stuff/) is still an improvement. I'm interested in learning more about this topic!

**Comment by swharden** [comment]
> a less fancy solution like csproj variables is still an improvement

If this is the case I think we really only need to use central versioning for:
* `ScottPlot` - the four csproj files linked in #1311
* `Avalonia`, `Avalonia.Desktop`, and `Avalonia.Diagnostics`
* `System.Drawing.Common` (keep at 4.6.1)
  * related: #1004
  * related: #1005

The rest can remain hard-coded in the csproj files.

**Comment by Kritner** [comment]
>> I can't seem to build all of the projects, mostly around the pieces that require framework and/or windows based functionality
>
> This might make this difficult 😅
>
> We exert a lot of effort to maintain cross-targeted builds across .NET Framework (back to 4.6.1), .NET Standard (2.0), and .NET 5 so the library runs on Windows, MacOS, and Linux.

The only sandbox app I can't build is "WinFormsFrameworkApp".  Here are outputs from the other demo apps:

WPF:
image

WinForms:
image

Console:
image

Avalonia Demo:
image

-----

FWIW, I ran `dotnet format` but that seems to be failing the CI / Check Formatting step. Additionally all tests are passing locally, I'll look at which one failed in that part of the process.

-----

> * System.Drawing.Common (keep at 4.6.1)
>   * related: Which version of System.Drawing.Common should ScottPlot depend on? #1004
>   * related: System.Drawing.Common version downgrade #1005

Addressed in https://github.com/ScottPlot/ScottPlot/pull/1310/commits/2ccae4f57a68515844efdfadbd79221a73bc01c0

Puts the central version to 4.6.1, individual packages can be upgraded with `OverrideVersion` in their respective csproj file if desired.  (I'm assuming no newer features were actually required, so I just kept everything at 4.6.1)

**Comment by swharden** [comment]
> @swharden should the package version be revved within this PR to *.25? or is that a part of the build process and/or rollup of several features into a new version?

Let's leave all the ScottPlot versions what they are. After merging this PR I'll automate the process of bumping them using GitHub Actions 👍

**Comment by Kritner** [comment]
> > @swharden should the package version be revved within this PR to *.25? or is that a part of the build process and/or rollup of several features into a new version?
> 
> Let's leave all the ScottPlot versions what they are. After merging this PR I'll automate the process of bumping them using GitHub Actions 👍

Sounds good, current branch is still set at version 4.1.24.  I *think* that's everything, except for the CI / Check Formatting step (which I did run the formatting...) and The failing Test on Linux: https://github.com/ScottPlot/ScottPlot/pull/1310/checks?check_run_id=3773009138#step:6:23

**Comment by swharden** [comment]
> The only sandbox app I can't build is "WinFormsFrameworkApp". Here are outputs from the other demo apps:

This is encouraging! As far as building/testing I'm happy to learn on the CI system. 

### Test_LettersDontRenderAsRectangles_SerifFont

Thanks again for your work on this! Sorry it got complicated by this 😝 

☠️ It looks like the Linux tests are failing due to font rendering issues. This may be because GitHub (the system that runs on the cloud) just today changed their default supported font set. To get this PR un-stuck you can either:

* Option 1: edit the [test that fails on Linux](https://github.com/ScottPlot/ScottPlot/blob/733441bc3878ba831d0d855575f8ced8515769d4/src/tests/Misc/Drawing.cs#L153-L179) and add a statement something like:

```cs
if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) Assert.Pass();
```

* Option 2: In [ci.yaml](https://github.com/ScottPlot/ScottPlot/blob/master/.github/workflows/ci.yaml) add an indented line after `strategy:` with `fail-fast: false` to allow the Windows tests to run even if the Linux tests fail.

**Comment by swharden** [comment]
## Auto-Formatting

Sometimes the formatter requires updating. Running this in the top-level solution directory (`src`) may solve your problem. If not let me know and I can autoformat it from my end 👍 

```bash
dotnet tool update -g dotnet-format
dotnet format
```

**Comment by Kritner** [comment]
> ## Auto-Formatting
> Sometimes the formatter requires updating. Running this in the top-level solution directory (`src`) may solve your problem. If not let me know and I can autoformat it from my end 👍
> 
> ```shell
> dotnet tool update -g dotnet-format
> dotnet format
> ```

Yeah might need you to give it a whirl, I did run it locally, which caused changes.

I don't know if it *was* those changes causing the step to fail or not, but `dotnet format` definitely changed those files as a part of one of my commits, specifically https://github.com/ScottPlot/ScottPlot/pull/1310/commits/1a7586116d8689ca5ea5bb44eac445d0ca9d89ba

**Comment by swharden** [comment]
I think the new errors indicate the package versioning isn't working as expected 😅 

The font rendering error was also a result of this actually... if the wrong version of System.Drawing.Common is installed on Linux, characters render as squares, and that test fails: https://github.com/ScottPlot/ScottPlot/issues/1079#issuecomment-849936525

Let's ignore autoformatting for now

File: Directory.Build.props
Comment (bclehmann) [inline Directory.Build.props:7]: I think this should be `0.10.*`

File: src/tests/Misc/Drawing.cs
Comment (bclehmann) [inline src/tests/Misc/Drawing.cs]: This is dicey, but I think if we have to not run it on Linux we should probably do this instead:

```cs
if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
```

**Review by bclehmann — APPROVED** [review]
Looks good to me, although that test does bug me. Especially as Linux font rendering on GDI+ seems to be more fragile than other OSes, and we're already rather dependent on bug reports to catch Linux issues.

File: .github/workflows/ci.yaml
Comment (bclehmann) [reply .github/workflows/ci.yaml:50]: I think now that the Linux test has been fixed this can be removed?

Comment (swharden) [reply .github/workflows/ci.yaml:50]: if `fail-fast` is `true` (default) then the failure of one job terminates all other jobs (so we never find out if they would have succeeded).

Setting this to `false` lets us wait-out the result of the Windows tests to see if those pass or fail, even though the Linux tests failed quickly. A failure in any case blocks the pipeline from proceeding

https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast

File: Directory.Build.props
Comment (swharden) [reply Directory.Build.props:7]:
```
error NU1011: Centrally defined floating package versions are not allowed
```

After fighting with `ManagePackageVersionsCentrally` for a few hours now, I think this is the last straw 😅 

I'll try not to get enticed by preview features in the future. I'll rework this PR to focus on improved ScottPlot version bumping and not worry about centralizing versions of dependencies.

I bet we can achieve the `*` behavior with a shared project variable...

File: src/tests/Misc/Drawing.cs
Comment (swharden) [reply src/tests/Misc/Drawing.cs]: Good catch! However I'll revert this test to its original form and, now that the correct version of System.Drawing.Common is loaded, the test passes again on all operating systems.

Comment (swharden) [reply src/tests/Misc/Drawing.cs]: c34cd0c

**Review by swharden — APPROVED** [review]

**Comment by swharden** [comment]
Wow, this was a rollercoaster ride! 🎢🎃 

## I did not side with centralized dependency versioning

* It couples packages more than I desire - the solution file is used for the convenience of the developer, but packages should be largely self-contained and isolated from the rest of the stuff in this repository.

* Mixed projects are not supported - I wish I could centrally version just `System.Drawing.Common` but it doesn't look like that's possible without centrally versioning everything, or requiring building from the _solution_ (I want the nuget packages to be independently buildable because it's often hard to build mixed-target solutions on cloud platforms)

* Wildcard versioning is currently unsupported. Allowing a version like `0.10.*` is pretty helpful.

I learned a lot through this experience, including not to rush to adopt preview features 😅

I moved the .props files to /dev/ so we might be able to use them later when the .NET tools mature

## The CI pipeline improved

* Tail-fast is now disabled, allowing more thorough assessments of whether failing tests fail on all operating systems or not

* The nuget packages are built from separate csproj files than the ones used by the solution file, so a new job was added to test building those and storing them as artifacts so you can download/inspect/test them. They're only retained for 24 hours to keep storage utilization low.

## Conditional dependency on `System.Drawing.Common`

Unless I goofed and haven't realized it yet, I think this is a big win of this PR. The System.Drawing.Common versioning problems (#1004) are limited to what happens when .NET Framework users _downgrade_ their package version. Before this PR we just targeted the lowest version, but that causes poor font rendering on linux (#1079). This new discussion led to the following in the nuget package csproj files, which I think solves the problem by allowing the latest version to be distributed to everyone except .NET Framework users (though they still have the ability to upgrade on their end if they wish).

https://github.com/ScottPlot/ScottPlot/blob/4895957f2507465017b3a1148066df12b1727324/src/controls/ScottPlot.WinForms/ScottPlot.WinForms.NUGET.csproj#L38-L47

**Pull request merged**
